contracts/sui/xcall/sources/main.move

```diff
@@ -0,0 +1,547 @@
+    use xcall::envelope::{Self,XCallEnvelope};
+    use xcall::connections::{Self,register};
+    use xcall::message_request::{Self, CSMessageRequest};
+    use xcall::call_message::{Self};
+    use xcall::persistent_message::{Self};
+    use xcall::call_message_rollback::{Self,CallMessageWithRollback};
+    use xcall::message_result::{Self,CSMessageResult};
+    use xcall::cs_message::{Self};
+    use xcall::rollback_data::{Self,RollbackData};
+    use xcall::xcall_state::{Self,Storage,AdminCap,IDCap,ConnCap};
+    use xcall::execute_ticket::{Self,ExecuteTicket};
+    use xcall::rollback_ticket::{Self,RollbackTicket};
+    use xcall::cs_message::{CSMessage};
+    use sui::bag::{Bag, Self};
+    use sui::table::{Table,Self};
+    use sui::package::{Self,Publisher};
+  
+    use sui::vec_map::{Self, VecMap};
+    use sui::versioned::{Self, Versioned};
+   
+     use sui::address::{Self};
+
+
+    const ENotOneTimeWitness: u64 = 0;
+    const ENotAdmin: u64 = 1;
+    const ENotUpgrade: u64 = 2;
+    const EWrongVersion: u64 = 3;
+    const EInvalidNID: u64 = 4;
+    const EInvalidSource: u64 = 5;
+    const ENoRollback: u64 = 6;
+    const EInvalidReply: u64 = 7;
+    const EDataMismatch: u64 = 8;
+    const EInvalidMsgType: u64 = 9;
+    const ERollbackNotEnabled:u64 = 10;
+    const EInfallible:u64 = 11;
+    const EInvalidAccess:u64 =12;
+
+    const MAX_DATA_SIZE: u64 = 2048;
+    const CURRENT_VERSION: u64 = 1;
+
+    /*************Events*****************/
+
+    public struct CallMessageSent has copy, drop{
+        from:String,
+        to:String,
+        sn:u128,
+    }
+
+    public struct CallMessage has copy, drop{
+        from:NetworkAddress,
+        to:String,
+        sn:u128,
+        req_id:u128,
+        data:vector<u8>,
+    }
+
+    public struct CallExecuted has copy, drop{
+        req_id:u128,
+        code:u8,
+        err_msg: String
+    }
+
+    public struct RollbackMessage has copy, drop{
+        sn:u128,
+        dapp: String,
+        data:vector<u8>,
+    }
+
+    public struct RollbackExecuted has copy, drop{
+        sn:u128
+    }
+
+    public struct ResponseMessage has copy, drop{
+        sn:u128,
+        response_code: u8
+    }
+    /***************/
+   
+    
+    fun init(ctx: &mut TxContext) {
+        let admin = xcall_state::create_admin_cap(ctx);
+        let storage = xcall_state::create_storage(
+           CURRENT_VERSION,
+            &admin,
+            ctx
+        );
+       xcall_state::share(storage);
+       xcall_state::transfer_admin_cap(admin,ctx);
+    }
+
+    public fun register_dapp<T: drop>(self:&Storage,
+        witness: T,
+        ctx: &mut TxContext
+    ):IDCap {
+        self.create_id_cap(ctx)
+    }
+
+    entry public fun configure_nid(self:&mut Storage,admin:&AdminCap, net_id:String,ctx: &mut TxContext){
+        if(!(net_id == b"".to_string())){
+            xcall_state::set_net_id(self,net_id);
+        }
+    }
```

Comment:
yeah, we cannot have this in constructor, so this function is needed, and providing AdminCap is the proof that it is called by admin

Reply:
Hmm okey, I also see than SUI only uses a ADMIN and not a Owner + Admin model. Any reason for that change? Since Admin should not have access to change stuff like this which could result in a hack or similar.
Owner should manage upgrades + very dangerous functions but set_protocol_fee or set feeHandler can be Admin

Reply:
ya sui doesnt have concrete owner concept , but deployer can be considered owner and its address can be stored or an ownercap issued to deployer if we want clear distinction between admin and owner. The deployer then can transfer adminCap to other account.